## DAGMA - EDA

### Este es el que usaremos para situacion 2, la situacion 1 ya quedo lista

In [315]:
import pandas as pd
import unicodedata
import numpy as np

In [316]:
df = pd.read_parquet("dagma/dagma_cvc_horario_raw.parquet")

In [317]:
df.head()

,estacion_id,nombre_est,nombre_fgda,msfl_code,med_concentracion_estandar,med_fecha_inicio,med_fecha_final,nombre_unidad,sigla_unidad,latitud,longitud,altitud,municipio,departamento
0,8285,BASE AÉREA,DAGMA,O3,2.84,2020-01-21 15:00:00,2020-01-21 16:00:00,Microgramos por metro cúbico,ug/m3,3.457128,-76.502303,956,Santiago de Cali,Valle del Cauca
1,8285,BASE AÉREA,DAGMA,O3,1.20,2020-01-21 20:00:00,2020-01-21 21:00:00,Microgramos por metro cúbico,ug/m3,3.457128,-76.502303,956,Santiago de Cali,Valle del Cauca
2,8285,BASE AÉREA,DAGMA,O3,0.00,2020-01-22 00:00:00,2020-01-22 01:00:00,Microgramos por metro cúbico,ug/m3,3.457128,-76.502303,956,Santiago de Cali,Valle del Cauca
3,8285,BASE AÉREA,DAGMA,O3,0.00,2020-01-22 01:00:00,2020-01-22 02:00:00,Microgramos por metro cúbico,ug/m3,3.457128,-76.502303,956,Santiago de Cali,Valle del Cauca
4,8285,BASE AÉREA,DAGMA,O3,0.10,2020-01-22 02:00:00,2020-01-22 03:00:00,Microgramos por metro cúbico,ug/m3,3.457128,-76.502303,956,Santiago de Cali,Valle del Cauca


In [318]:
df.duplicated().sum()

np.int64(0)

In [319]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 107291 entries, 0 to 107290
Data columns (total 14 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   estacion_id                 107291 non-null  str           
 1   nombre_est                  107291 non-null  str           
 2   nombre_fgda                 107291 non-null  str           
 3   msfl_code                   107291 non-null  str           
 4   med_concentracion_estandar  107291 non-null  float64       
 5   med_fecha_inicio            107291 non-null  datetime64[us]
 6   med_fecha_final             107291 non-null  datetime64[us]
 7   nombre_unidad               107291 non-null  str           
 8   sigla_unidad                107291 non-null  str           
 9   latitud                     107291 non-null  float64       
 10  longitud                    107291 non-null  float64       
 11  altitud                     107291 non-null  int64

In [320]:
df.rename(columns={'med_fecha_inicio': 'fecha'}, inplace=True)

In [321]:
df = df.drop(columns=["departamento", "nombre_unidad", "sigla_unidad", "med_fecha_final"])

In [322]:
df["municipio"].unique()

<ArrowStringArray>
['Santiago de Cali', 'Yumbo']
Length: 2, dtype: str

In [323]:
df.describe()

,med_concentracion_estandar,fecha,latitud,longitud,altitud
count,107291.000000,107291,107291.000000,107291.000000,107291.000000
mean,10.794378,2021-10-06 16:55:46.166966,3.479262,-76.504944,961.585408
min,0.000000,2020-01-01 00:00:00,3.304517,-76.549613,950.000000
25%,1.620000,2020-08-07 05:00:00,3.428260,-76.518058,950.000000
50%,5.470000,2021-03-25 08:00:00,3.457128,-76.502303,956.000000
75%,11.840000,2022-08-21 08:00:00,3.579075,-76.489558,968.000000
max,241.190000,2024-12-31 23:00:00,3.579075,-76.466584,994.000000
std,16.832756,NaN,0.079095,0.020942,14.185082


In [324]:
df.head()

,estacion_id,nombre_est,nombre_fgda,msfl_code,med_concentracion_estandar,fecha,latitud,longitud,altitud,municipio
0,8285,BASE AÉREA,DAGMA,O3,2.84,2020-01-21 15:00:00,3.457128,-76.502303,956,Santiago de Cali
1,8285,BASE AÉREA,DAGMA,O3,1.20,2020-01-21 20:00:00,3.457128,-76.502303,956,Santiago de Cali
2,8285,BASE AÉREA,DAGMA,O3,0.00,2020-01-22 00:00:00,3.457128,-76.502303,956,Santiago de Cali
3,8285,BASE AÉREA,DAGMA,O3,0.00,2020-01-22 01:00:00,3.457128,-76.502303,956,Santiago de Cali
4,8285,BASE AÉREA,DAGMA,O3,0.10,2020-01-22 02:00:00,3.457128,-76.502303,956,Santiago de Cali


In [325]:
df = df.rename(columns={
                        'msfl_code': 'tipo_gas', 
                        'med_fecha': "fecha", 
                        'med_fechaal': "fecha", 
                        "med_concentracion_estandar":"concentracion",
                        "nombre_fgda": "fuente"
                    })

In [326]:
df.columns = df.columns.str.lower()
df.columns = df.columns.str.replace('´', '')

In [327]:
df.head()

,estacion_id,nombre_est,fuente,tipo_gas,concentracion,fecha,latitud,longitud,altitud,municipio
0,8285,BASE AÉREA,DAGMA,O3,2.84,2020-01-21 15:00:00,3.457128,-76.502303,956,Santiago de Cali
1,8285,BASE AÉREA,DAGMA,O3,1.20,2020-01-21 20:00:00,3.457128,-76.502303,956,Santiago de Cali
2,8285,BASE AÉREA,DAGMA,O3,0.00,2020-01-22 00:00:00,3.457128,-76.502303,956,Santiago de Cali
3,8285,BASE AÉREA,DAGMA,O3,0.00,2020-01-22 01:00:00,3.457128,-76.502303,956,Santiago de Cali
4,8285,BASE AÉREA,DAGMA,O3,0.10,2020-01-22 02:00:00,3.457128,-76.502303,956,Santiago de Cali


In [328]:
def limpiar_texto(texto):
    if isinstance(texto, str):
        texto = texto.lower()
        texto = unicodedata.normalize('NFKD', texto).encode('ASCII', 'ignore').decode('utf-8')
        return texto
    return texto

In [329]:
columnas_str = df.select_dtypes(include=['object', 'string']).columns
df[columnas_str] = df[columnas_str].map(limpiar_texto)

In [330]:
df.head()

,estacion_id,nombre_est,fuente,tipo_gas,concentracion,fecha,latitud,longitud,altitud,municipio
0,8285,base aerea,dagma,o3,2.84,2020-01-21 15:00:00,3.457128,-76.502303,956,santiago de cali
1,8285,base aerea,dagma,o3,1.20,2020-01-21 20:00:00,3.457128,-76.502303,956,santiago de cali
2,8285,base aerea,dagma,o3,0.00,2020-01-22 00:00:00,3.457128,-76.502303,956,santiago de cali
3,8285,base aerea,dagma,o3,0.00,2020-01-22 01:00:00,3.457128,-76.502303,956,santiago de cali
4,8285,base aerea,dagma,o3,0.10,2020-01-22 02:00:00,3.457128,-76.502303,956,santiago de cali


In [331]:
print((df['concentracion'] == 0).sum())

12206


In [332]:
df["fecha"].dt.year.value_counts().sort_index()

fecha
2020    52192
2021    11514
2022    21569
2023     9825
2024    12191
Name: count, dtype: int64

In [333]:
gases_por_estacion = df.groupby('nombre_est')['tipo_gas'].unique()
print(gases_por_estacion)

nombre_est
base aerea                    [o3, so2]
canaveralejo                      [so2]
compartir                          [o3]
era obrero                         [o3]
estacion yumbo           [no2, o3, so2]
la ermita                         [so2]
la flora                      [o3, so2]
pance                              [o3]
transitoria-navarro           [o3, so2]
universidad del valle              [o3]
Name: tipo_gas, dtype: object


In [334]:
rango_fechas = df.groupby('nombre_est').agg(
    Fecha_Minima=('fecha', 'min'),
    Fecha_Maxima=('fecha', 'max'),
    Total_Registros=('nombre_est', 'size')
).reset_index()

print("--- Rango de fechas por cada estación ---")
print(rango_fechas)

--- Rango de fechas por cada estación ---
              nombre_est        Fecha_Minima        Fecha_Maxima  \
0             base aerea 2020-01-07 12:00:00 2023-09-14 23:00:00   
1           canaveralejo 2020-01-01 00:00:00 2022-06-20 13:00:00   
2              compartir 2020-01-01 00:00:00 2022-12-31 23:00:00   
3             era obrero 2020-01-01 00:00:00 2022-10-25 23:00:00   
4         estacion yumbo 2020-07-01 01:00:00 2024-12-31 23:00:00   
5              la ermita 2020-01-01 00:00:00 2022-10-25 23:00:00   
6               la flora 2020-01-01 00:00:00 2022-09-14 10:00:00   
7                  pance 2020-01-01 00:00:00 2022-10-27 23:00:00   
8    transitoria-navarro 2020-01-28 11:00:00 2020-12-07 08:00:00   
9  universidad del valle 2020-05-03 06:00:00 2023-04-14 08:00:00   

   Total_Registros  
0            15080  
1             5053  
2             7507  
3             6939  
4            34372  
5             8627  
6            10067  
7             5861  
8             7435  

transitoria navarro tiene solo datos del 2020

nosotros eliminamos 2020 porque nuestro panel va de 2021 al 2025

In [335]:
df = df[df['fecha'].dt.year != 2020]

In [336]:
df["fecha"].dt.year.value_counts()

fecha
2022    21569
2024    12191
2021    11514
2023     9825
Name: count, dtype: int64

In [337]:
df.head(20)

,estacion_id,nombre_est,fuente,tipo_gas,concentracion,fecha,latitud,longitud,altitud,municipio
3183,8285,base aerea,dagma,o3,0.73,2021-07-09 22:00:00,3.457128,-76.502303,956,santiago de cali
3184,8285,base aerea,dagma,o3,0.80,2021-07-10 00:00:00,3.457128,-76.502303,956,santiago de cali
3185,8285,base aerea,dagma,o3,0.63,2021-07-10 01:00:00,3.457128,-76.502303,956,santiago de cali
3186,8285,base aerea,dagma,o3,0.80,2021-07-10 03:00:00,3.457128,-76.502303,956,santiago de cali
3187,8285,base aerea,dagma,o3,0.82,2021-07-10 04:00:00,3.457128,-76.502303,956,santiago de cali
3188,8285,base aerea,dagma,o3,0.37,2021-08-20 00:00:00,3.457128,-76.502303,956,santiago de cali
3189,8285,base aerea,dagma,o3,1.28,2021-08-20 01:00:00,3.457128,-76.502303,956,santiago de cali
3190,8285,base aerea,dagma,o3,1.67,2021-08-20 02:00:00,3.457128,-76.502303,956,santiago de cali
3191,8285,base aerea,dagma,o3,0.57,2021-08-20 03:00:00,3.457128,-76.502303,956,santiago de cali
3192,8285,base aerea,dagma,o3,0.69,2021-08-20 04:00:00,3.457128,-76.502303,956,santiago de cali


In [338]:
df.info()

<class 'pandas.DataFrame'>
Index: 55099 entries, 3183 to 107290
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   estacion_id    55099 non-null  str           
 1   nombre_est     55099 non-null  str           
 2   fuente         55099 non-null  str           
 3   tipo_gas       55099 non-null  str           
 4   concentracion  55099 non-null  float64       
 5   fecha          55099 non-null  datetime64[us]
 6   latitud        55099 non-null  float64       
 7   longitud       55099 non-null  float64       
 8   altitud        55099 non-null  int64         
 9   municipio      55099 non-null  str           
dtypes: datetime64[us](1), float64(3), int64(1), str(5)
memory usage: 6.4 MB


aqui ya tenemos solo 55099 datos para usar ya que solo usamos de 2021 al 2026.

Y DAGMA viene del 2020 al 2024

In [339]:
print((df['concentracion'] == 0).sum())

2006


In [340]:
print(df[df['concentracion'] == 0][['fecha', 'concentracion']])

                     fecha  concentracion
3301   2021-08-27 01:00:00            0.0
3433   2021-09-03 00:00:00            0.0
3434   2021-09-03 01:00:00            0.0
3502   2021-09-06 23:00:00            0.0
3618   2021-09-13 07:00:00            0.0
...                    ...            ...
106244 2022-08-07 09:00:00            0.0
106292 2022-08-09 18:00:00            0.0
106331 2022-08-11 17:00:00            0.0
106376 2022-08-13 17:00:00            0.0
106421 2022-08-15 18:00:00            0.0

[2006 rows x 2 columns]
                     fecha  concentracion
3301   2021-08-27 01:00:00            0.0
3433   2021-09-03 00:00:00            0.0
3434   2021-09-03 01:00:00            0.0
3502   2021-09-06 23:00:00            0.0
3618   2021-09-13 07:00:00            0.0
...                    ...            ...
106244 2022-08-07 09:00:00            0.0
106292 2022-08-09 18:00:00            0.0
106331 2022-08-11 17:00:00            0.0
106376 2022-08-13 17:00:00            0.0
106421 20

In [341]:
df['fecha'] = pd.to_datetime(df['fecha'])

ceros = df[df['concentracion'] == 0]
ceros_por_hora = ceros.groupby(ceros['fecha'].dt.hour).size()

print("Cantidad de ceros por hora del día:")
print(ceros_por_hora)

Cantidad de ceros por hora del día:
fecha
0      52
1      53
2      61
3      62
4      78
5      84
6      72
7      34
8      23
9      12
10     15
11     26
12     51
13     74
14    119
15    143
16    159
17    161
18    162
19    152
20    137
21    115
22     92
23     69
dtype: int64


In [342]:
ceros_por_estacion = ceros.groupby(['nombre_est', 'tipo_gas']).size()

print("Ceros totales por Estación y Gas:")
print(ceros_por_estacion)

Ceros totales por Estación y Gas:
nombre_est             tipo_gas
base aerea             o3            17
era obrero             o3             1
estacion yumbo         no2           64
                       o3           215
                       so2         1680
la ermita              so2           17
la flora               o3             5
universidad del valle  o3             7
dtype: int64


In [343]:
porcentaje_ceros = (df['concentracion'] == 0).groupby(df['nombre_est']).mean() * 100

print("Porcentaje de ceros por estación:")
print(porcentaje_ceros.map("{:.2f}%".format))

Porcentaje de ceros por estación:
nombre_est
base aerea               0.23%
canaveralejo             0.00%
compartir                0.00%
era obrero               0.05%
estacion yumbo           6.58%
la ermita                0.61%
la flora                 0.32%
pance                    0.00%
universidad del valle    0.16%
Name: concentracion, dtype: str


In [344]:
gases = ceros.groupby(['nombre_est', 'tipo_gas']).size()

print("Ceros totales por Estación y Gas:")
print(gases)

Ceros totales por Estación y Gas:
nombre_est             tipo_gas
base aerea             o3            17
era obrero             o3             1
estacion yumbo         no2           64
                       o3           215
                       so2         1680
la ermita              so2           17
la flora               o3             5
universidad del valle  o3             7
dtype: int64


In [345]:
gases_por_estacion = df.groupby('nombre_est')['tipo_gas'].unique()
print(gases_por_estacion)

nombre_est
base aerea                    [o3, so2]
canaveralejo                      [so2]
compartir                          [o3]
era obrero                         [o3]
estacion yumbo           [no2, o3, so2]
la ermita                         [so2]
la flora                           [o3]
pance                              [o3]
universidad del valle              [o3]
Name: tipo_gas, dtype: object


In [346]:
rango_fechas = df.groupby('nombre_est').agg(
    Fecha_Minima=('fecha', 'min'),
    Fecha_Maxima=('fecha', 'max'),
    Total_Registros=('nombre_est', 'size') # Opcional: para saber cuántas horas/filas tiene cada una
).reset_index()

print("--- Rango de fechas por cada estación ---")
print(rango_fechas)

--- Rango de fechas por cada estación ---
              nombre_est        Fecha_Minima        Fecha_Maxima  \
0             base aerea 2021-07-09 22:00:00 2023-09-14 23:00:00   
1           canaveralejo 2021-11-16 19:00:00 2022-06-20 13:00:00   
2              compartir 2021-08-18 13:00:00 2022-12-31 23:00:00   
3             era obrero 2021-10-13 00:00:00 2022-10-25 23:00:00   
4         estacion yumbo 2021-01-01 00:00:00 2024-12-31 23:00:00   
5              la ermita 2021-10-27 09:00:00 2022-10-25 23:00:00   
6               la flora 2022-04-13 09:00:00 2022-09-14 10:00:00   
7                  pance 2021-11-24 21:00:00 2022-10-27 23:00:00   
8  universidad del valle 2021-08-26 17:00:00 2023-04-14 08:00:00   

   Total_Registros  
0             7356  
1             1678  
2             3383  
3             1917  
4            29776  
5             2796  
6             1575  
7             2369  
8             4249  


Aqui vemos que hay un problema con el tema de que solo Yumbo reporta NO2 y el resto pailas. Solo reportan el resto de gases

In [347]:
concentracion_mayor_cero = (df['concentracion'] > 0).groupby(df['nombre_est']).mean() * 100

print("Porcentaje de concentraciones mayores a cero por estación:")
print(concentracion_mayor_cero.map("{:.2f}%".format))

Porcentaje de concentraciones mayores a cero por estación:
nombre_est
base aerea                99.77%
canaveralejo             100.00%
compartir                100.00%
era obrero                99.95%
estacion yumbo            93.42%
la ermita                 99.39%
la flora                  99.68%
pance                    100.00%
universidad del valle     99.84%
Name: concentracion, dtype: str


Lo bueno es que la mayoria estan bien monitoreados

In [360]:
df.info()

<class 'pandas.DataFrame'>
Index: 55099 entries, 3183 to 107290
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   estacion_id    55099 non-null  str           
 1   nombre_est     55099 non-null  str           
 2   fuente         55099 non-null  str           
 3   tipo_gas       55099 non-null  str           
 4   concentracion  55099 non-null  float64       
 5   fecha          55099 non-null  datetime64[us]
 6   latitud        55099 non-null  float64       
 7   longitud       55099 non-null  float64       
 8   altitud        55099 non-null  int64         
 9   municipio      55099 non-null  str           
dtypes: datetime64[us](1), float64(3), int64(1), str(5)
memory usage: 6.3 MB


## DAGMA CRUCE CON EL DE CRISTIAN

In [348]:
df_raw = pd.read_excel('dagma/dagma-cristian.xlsx', header=None)

fila_estaciones = df_raw.iloc[22].ffill().astype(str).tolist()
fila_gases = df_raw.iloc[23].ffill().astype(str).tolist()

nuevas_columnas = []
for est, gas in zip(fila_estaciones, fila_gases):
    if "fecha" in est.lower() or "hora" in est.lower():
        nuevas_columnas.append("fecha")
    else:
        nuevas_columnas.append(f"{est.strip()}||{gas.strip()}")

df_wide = df_raw.iloc[25:].copy()
df_wide.columns = nuevas_columnas

df_wide = df_wide.replace(['NA', 'nan', 'NaN', ' '], np.nan)

df_vertical = pd.melt(
    df_wide, 
    id_vars=['fecha'],                 
    var_name='estacion_y_gas',          
    value_name='concentracion'          
)


df_vertical[['nombre_est', 'tipo_gas']] = df_vertical['estacion_y_gas'].str.split('||', expand=True, regex=False)

df_vertical = df_vertical.drop(columns=['estacion_y_gas'])

df_vertical['fecha'] = df_vertical['fecha'].astype(str)
df_vertical['fecha'] = df_vertical['fecha'].str.replace(' 24:00', ' 00:00', regex=False)
df_vertical['fecha'] = pd.to_datetime(df_vertical['fecha'], format='mixed')

df_vertical = df_vertical[['nombre_est', 'tipo_gas', 'concentracion', 'fecha']]

print(df_vertical.head(10))

   nombre_est tipo_gas concentracion               fecha
0  Base Aerea     PM25           158 2020-01-01 01:00:00
1  Base Aerea     PM25            51 2020-01-01 02:00:00
2  Base Aerea     PM25            44 2020-01-01 03:00:00
3  Base Aerea     PM25            42 2020-01-01 04:00:00
4  Base Aerea     PM25            44 2020-01-01 05:00:00
5  Base Aerea     PM25            31 2020-01-01 06:00:00
6  Base Aerea     PM25            30 2020-01-01 07:00:00
7  Base Aerea     PM25            57 2020-01-01 08:00:00
8  Base Aerea     PM25            47 2020-01-01 09:00:00
9  Base Aerea     PM25            54 2020-01-01 10:00:00


In [349]:
df_cristian = df_vertical.copy()

df_cristian.head()

,nombre_est,tipo_gas,concentracion,fecha
0,Base Aerea,PM25,158,2020-01-01 01:00:00
1,Base Aerea,PM25,51,2020-01-01 02:00:00
2,Base Aerea,PM25,44,2020-01-01 03:00:00
3,Base Aerea,PM25,42,2020-01-01 04:00:00
4,Base Aerea,PM25,44,2020-01-01 05:00:00


In [350]:
df_cristian["tipo_gas"].unique()

<ArrowStringArray>
[                'PM25',                 'PM10',         'Black Carbon',
                'UV-PM',                   'O3',                  'SO2',
                  'H2S',                  'NO2',          'Temperatura',
              'Humedad',           'Vel Viento',           'Dir Viento',
               'Lluvia',      'Radiacion Solar',      'Presion Baromet',
 'Presion Baromet _NEF']
Length: 16, dtype: str

In [351]:
lo_que_me_sirve = ['NO2', 'SO2', 'O3']
df_cristian= df_cristian[df_cristian['tipo_gas'].isin(lo_que_me_sirve)]

In [352]:
df_cristian['tipo_gas'].unique()

<ArrowStringArray>
['O3', 'SO2', 'NO2']
Length: 3, dtype: str

In [353]:
df_cristian.info()

<class 'pandas.DataFrame'>
Index: 736512 entries, 999552 to 1893887
Data columns (total 4 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   nombre_est     736512 non-null  str           
 1   tipo_gas       736512 non-null  str           
 2   concentracion  414216 non-null  object        
 3   fecha          736512 non-null  datetime64[us]
dtypes: datetime64[us](1), object(1), str(2)
memory usage: 36.9+ MB


In [354]:
columnas_str = df_cristian.select_dtypes(include=['object', 'string']).columns
df_cristian[columnas_str] = df_cristian[columnas_str].map(limpiar_texto)

In [355]:
print("DF CRISTIAN")
print(df_cristian['fecha'].unique(), "\n")
print(df_cristian['nombre_est'].unique(), "\n")
print(df_cristian['tipo_gas'].unique(), "\n")

print("DF")

print(df['fecha'].unique(), "\n")
print(df['nombre_est'].unique(), "\n")
print(df['tipo_gas'].unique(), "\n")

DF CRISTIAN
<DatetimeArray>
['2020-01-01 01:00:00', '2020-01-01 02:00:00', '2020-01-01 03:00:00',
 '2020-01-01 04:00:00', '2020-01-01 05:00:00', '2020-01-01 06:00:00',
 '2020-01-01 07:00:00', '2020-01-01 08:00:00', '2020-01-01 09:00:00',
 '2020-01-01 10:00:00',
 ...
 '2025-12-31 15:00:00', '2025-12-31 16:00:00', '2025-12-31 17:00:00',
 '2025-12-31 18:00:00', '2025-12-31 19:00:00', '2025-12-31 20:00:00',
 '2025-12-31 21:00:00', '2025-12-31 22:00:00', '2025-12-31 23:00:00',
 '2025-12-31 00:00:00']
Length: 52608, dtype: datetime64[us] 

<ArrowStringArray>
[           'base aerea',             'compartir',          'era - obrero',
                 'flora',                 'pance', 'transitoria - navarro',
              'univalle',          'canaveralejo',                'ermita']
Length: 9, dtype: str 

<ArrowStringArray>
['o3', 'so2', 'no2']
Length: 3, dtype: str 

DF
<DatetimeArray>
['2021-07-09 22:00:00', '2021-07-10 00:00:00', '2021-07-10 01:00:00',
 '2021-07-10 03:00:00', '2021-07-10 

In [358]:
diccionario_estaciones = {
    'universidad del valle': 'univalle',
    'la ermita': 'ermita',
    'la flora': 'flora',
    'era obrero': 'era - obrero'
}

In [359]:
df['nombre_est'] = df['nombre_est'].replace(diccionario_estaciones)

In [356]:
df_cristian.dropna(subset=['concentracion'], inplace=True)

In [364]:
print(df_cristian.info())
print()

print(df.info())

<class 'pandas.DataFrame'>
Index: 414216 entries, 999552 to 1885064
Data columns (total 4 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   nombre_est     414216 non-null  str           
 1   tipo_gas       414216 non-null  str           
 2   concentracion  414216 non-null  float64       
 3   fecha          414216 non-null  datetime64[us]
dtypes: datetime64[us](1), float64(1), str(2)
memory usage: 20.4 MB
None

<class 'pandas.DataFrame'>
Index: 55099 entries, 3183 to 107290
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   estacion_id    55099 non-null  str           
 1   nombre_est     55099 non-null  str           
 2   fuente         55099 non-null  str           
 3   tipo_gas       55099 non-null  str           
 4   concentracion  55099 non-null  float64       
 5   fecha          55099 non-null  datetime64[us]
 6   lati

In [366]:
print(df_cristian["nombre_est"].unique())

print(df["nombre_est"].unique())

<ArrowStringArray>
[           'base aerea',             'compartir',          'era - obrero',
                 'flora',                 'pance', 'transitoria - navarro',
              'univalle',          'canaveralejo',                'ermita']
Length: 9, dtype: str
<ArrowStringArray>
[    'base aerea',   'canaveralejo',      'compartir',   'era - obrero',
 'estacion yumbo',         'ermita',          'flora',          'pance',
       'univalle']
Length: 9, dtype: str


In [367]:
gases = ['no2', 'o3', 'so2']

for gas in gases:
    print(f"\n============= COMPARACIÓN PARA EL GAS: {gas.upper()} =============")
    
    # Sacamos los estadísticos de cada DF para ese gas específico
    stats_cristian = df_cristian[df_cristian['tipo_gas'] == gas]['concentracion'].describe()
    stats_dagma = df[df['tipo_gas'] == gas]['concentracion'].describe()
    
    # Los unimos en una sola tablita para comparar lado a lado
    comparativa = pd.DataFrame({'DF Cristian': stats_cristian, 'DF Dagma (Original)': stats_dagma})
    print(comparativa.round(2)) # Redondeamos a 2 decimales para que se vea limpio


============= COMPARACIÓN PARA EL GAS: NO2 =============
       DF Cristian  DF Dagma (Original)
count     26259.00              4985.00
mean         14.06                10.73
std          10.05                10.18
min           0.00                 0.00
25%           6.96                 4.76
50%          12.15                 9.10
75%          18.97                13.60
max          91.86               163.90

============= COMPARACIÓN PARA EL GAS: O3 =============
       DF Cristian  DF Dagma (Original)
count    230761.00             30376.00
mean         28.06                17.00
std          31.20                22.23
min           0.00                 0.00
25%           6.89                 3.85
50%          14.87                 8.16
75%          38.84                20.50
max         404.18               194.90

============= COMPARACIÓN PARA EL GAS: SO2 =============
       DF Cristian  DF Dagma (Original)
count    157196.00             19738.00
mean          5.91         

In [368]:
df_cristian.loc[df_cristian['concentracion'] < 0, 'concentracion'] = 0

In [369]:
gases = ['no2', 'o3', 'so2']

for gas in gases:
    print(f"\n============= COMPARACIÓN PARA EL GAS: {gas.upper()} =============")
    stats_cristian = df_cristian[df_cristian['tipo_gas'] == gas]['concentracion'].describe()
    stats_dagma = df[df['tipo_gas'] == gas]['concentracion'].describe()
    
    comparativa = pd.DataFrame({'DF Cristian': stats_cristian, 'DF Dagma (Original)': stats_dagma})
    print(comparativa.round(2))


============= COMPARACIÓN PARA EL GAS: NO2 =============
       DF Cristian  DF Dagma (Original)
count     26259.00              4985.00
mean         14.06                10.73
std          10.05                10.18
min           0.00                 0.00
25%           6.96                 4.76
50%          12.15                 9.10
75%          18.97                13.60
max          91.86               163.90

============= COMPARACIÓN PARA EL GAS: O3 =============
       DF Cristian  DF Dagma (Original)
count    230761.00             30376.00
mean         28.06                17.00
std          31.20                22.23
min           0.00                 0.00
25%           6.89                 3.85
50%          14.87                 8.16
75%          38.84                20.50
max         404.18               194.90

============= COMPARACIÓN PARA EL GAS: SO2 =============
       DF Cristian  DF Dagma (Original)
count    157196.00             19738.00
mean          5.91         

LIMPIAR OUTLIERS CRISTIAN

In [373]:
# Para NO2
q1_no2_c = df_cristian[df_cristian['tipo_gas'] == 'no2']['concentracion'].quantile(0.25)
q3_no2_c = df_cristian[df_cristian['tipo_gas'] == 'no2']['concentracion'].quantile(0.75)
iqr_no2_c = q3_no2_c - q1_no2_c
lim_no2_c = q3_no2_c + 1.5 * iqr_no2_c

# Para O3
q1_o3_c = df_cristian[df_cristian['tipo_gas'] == 'o3']['concentracion'].quantile(0.25)
q3_o3_c = df_cristian[df_cristian['tipo_gas'] == 'o3']['concentracion'].quantile(0.75)
iqr_o3_c = q3_o3_c - q1_o3_c
lim_o3_c = q3_o3_c + 1.5 * iqr_o3_c

# Para SO2
q1_so2_c = df_cristian[df_cristian['tipo_gas'] == 'so2']['concentracion'].quantile(0.25)
q3_so2_c = df_cristian[df_cristian['tipo_gas'] == 'so2']['concentracion'].quantile(0.75)
iqr_so2_c = q3_so2_c - q1_so2_c
lim_so2_c = q3_so2_c + 1.5 * iqr_so2_c

# Aplicar los filtros combinados en df_cristian
df_cristian = df_cristian[
    ((df_cristian['tipo_gas'] == 'no2') & (df_cristian['concentracion'] <= lim_no2_c)) |
    ((df_cristian['tipo_gas'] == 'o3') & (df_cristian['concentracion'] <= lim_o3_c)) |
    ((df_cristian['tipo_gas'] == 'so2') & (df_cristian['concentracion'] <= lim_so2_c))
].copy()

LIMPIAR OUTLIERS DAGMA

In [374]:
# Para NO2
q1_no2_d = df[df['tipo_gas'] == 'no2']['concentracion'].quantile(0.25)
q3_no2_d = df[df['tipo_gas'] == 'no2']['concentracion'].quantile(0.75)
iqr_no2_d = q3_no2_d - q1_no2_d
lim_no2_d = q3_no2_d + 1.5 * iqr_no2_d

# Para O3
q1_o3_d = df[df['tipo_gas'] == 'o3']['concentracion'].quantile(0.25)
q3_o3_d = df[df['tipo_gas'] == 'o3']['concentracion'].quantile(0.75)
iqr_o3_d = q3_o3_d - q1_o3_d
lim_o3_d = q3_o3_d + 1.5 * iqr_o3_d

# Para SO2
q1_so2_d = df[df['tipo_gas'] == 'so2']['concentracion'].quantile(0.25)
q3_so2_d = df[df['tipo_gas'] == 'so2']['concentracion'].quantile(0.75)
iqr_so2_d = q3_so2_d - q1_so2_d
lim_so2_d = q3_so2_d + 1.5 * iqr_so2_d

# Aplicar los filtros combinados en df
df = df[
    ((df['tipo_gas'] == 'no2') & (df['concentracion'] <= lim_no2_d)) |
    ((df['tipo_gas'] == 'o3') & (df['concentracion'] <= lim_o3_d)) |
    ((df['tipo_gas'] == 'so2') & (df['concentracion'] <= lim_so2_d))
].copy()

In [375]:
gases = ['no2', 'o3', 'so2']

for gas in gases:
    print(f"\n============= COMPARACIÓN PARA EL GAS: {gas.upper()} =============")
    stats_cristian = df_cristian[df_cristian['tipo_gas'] == gas]['concentracion'].describe()
    stats_dagma = df[df['tipo_gas'] == gas]['concentracion'].describe()
    
    comparativa = pd.DataFrame({'DF Cristian': stats_cristian, 'DF Dagma (Original)': stats_dagma})
    print(comparativa.round(2))


============= COMPARACIÓN PARA EL GAS: NO2 =============
       DF Cristian  DF Dagma (Original)
count     25393.00              4776.00
mean         13.00                 9.35
std           8.27                 5.92
min           0.00                 0.00
25%           6.77                 4.53
50%          11.83                 8.80
75%          18.05                12.90
max          36.99                26.80

============= COMPARACIÓN PARA EL GAS: O3 =============
       DF Cristian  DF Dagma (Original)
count    214405.00             27413.00
mean         21.74                10.94
std          21.24                10.39
min           0.00                 0.00
25%           6.40                 3.59
50%          13.46                 7.10
75%          30.55                14.90
max          86.77                45.46

============= COMPARACIÓN PARA EL GAS: SO2 =============
       DF Cristian  DF Dagma (Original)
count    146880.00             17691.00
mean          3.76         